### PR TITLE
shift up and down feat

### PR DESCRIPTION
**Pull request by petrican**
rel to : https://github.com/metasfresh/metasfresh/issues/12171

Demo in SO listing - https://www.loom.com/share/29371b9f53674f7c9ecf790d059006fc

Demo in modal - https://www.loom.com/share/197efe65d11f4d4390a4a185e1bde4ad

### CODE REVIEW

**Comment by petrican** [comment]
Screenshot 2021-12-16 at 09 10 29
Holding CTRL and selecting separate rows still works

File: frontend/src/components/table/Table.js [hunk 1]
```diff
@@ -174,14 +187,12 @@ export default class Table extends PureComponent {
     }
 
     switch (e.key) {
-      case 'ArrowDown': {
+      case ARROW_DOWN_KEY: {
         e.preventDefault();
 
         const { currentId, array } = this.getCurrentRowId();
```
Comment (teosarca) [inline frontend/src/components/table/Table.js:193]: provide the proper param. don't assume default param.

File: frontend/src/components/table/Table.js [hunk 2]
```diff
@@ -72,13 +73,15 @@ export default class Table extends PureComponent {
     this.tfoot = ref;
   };
 
-  getCurrentRowId = () => {
+  getCurrentRowId = (arrowOrientation = ARROW_DOWN_KEY) => {
```
Comment (teosarca) [inline frontend/src/components/table/Table.js:76]: avoid default params in a "private" function...

File: frontend/src/components/table/Table.js [hunk 3]
```diff
@@ -99,12 +102,24 @@ export default class Table extends PureComponent {
     return arrayIndex.slice(selectedArr[0], selectedArr[1] + 1);
   };
 
+  /**
+   * @summary Updates the start reference used for the multi selection when SHIFT + arrow up/down keys are pressed
+   * @param {*} currentId - the current index in the array of rows
+   */
+  updateMultiSelStart = (currentId) => {
+    this.multiSelStart =
+      !this.multiSelStart && this.multiSelStart !== 0
```
Comment (teosarca) [inline frontend/src/components/table/Table.js:111]: pls simply this logic 
and i suggest to inline the function.

File: frontend/src/components/table/Table.js [hunk 4]
```diff
@@ -191,19 +202,29 @@ export default class Table extends PureComponent {
...
               showSelectedIncludedView([array[currentId + 1]])
           );
+          this.multiSelStart = null;
         } else {
-          handleSelect(array[currentId + 1], false, idFocused);
+          this.updateMultiSelStart(currentId);
+
+          const downShiftSel = array.slice(
+            this.multiSelStart > 0 ? this.multiSelStart : 0,
+            currentId + 2
```
Comment (teosarca) [inline frontend/src/components/table/Table.js:211]: add a small `//` comment and explain why `+ 2`

File: frontend/src/components/table/Table.js [hunk 5]
```diff
@@ -191,19 +202,29 @@ export default class Table extends PureComponent {
...
+          );
         }
         break;
       }
-      case 'ArrowUp': {
+      case ARROW_UP_KEY: {
         e.preventDefault();
 
-        const { currentId, array } = this.getCurrentRowId();
+        const { currentId, array } = this.getCurrentRowId(ARROW_UP_KEY);
```
Comment (teosarca) [inline frontend/src/components/table/Table.js:225]: find a better name for `array`

File: frontend/src/components/table/Table.js [hunk 6]
```diff
@@ -72,13 +73,15 @@ export default class Table extends PureComponent {
     this.tfoot = ref;
   };
 
-  getCurrentRowId = () => {
+  getCurrentRowId = (arrowOrientation = ARROW_DOWN_KEY) => {
     const { keyProperty, selected, rows } = this.props;
 
     const array = rows.map((item) => item[keyProperty]);
```
Comment (teosarca) [inline frontend/src/components/table/Table.js:79]: find a better name for this `array`.
Maybe `selectedRowIds` ? (not sure what it does)

File: frontend/src/components/table/Table.js [hunk 7]
```diff
@@ -20,6 +20,7 @@ export default class Table extends PureComponent {
       listenOnKeys: true,
       tableRefreshToggle: false,
     };
+    this.multiSelStart = null;
```
Comment (teosarca) [inline frontend/src/components/table/Table.js:23]: find a better name for `multiSelStart`.
Maybe something like `multiSelectionStartId`? (or something which when i read I know what's that).

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

File: frontend/src/components/table/Table.js (hunk 1 above)
Comment (petrican) [reply frontend/src/components/table/Table.js:193]: ok fixing to 

const { currentId, array } = this.getCurrentRowId(ARROW_DOWN_KEY);

File: frontend/src/components/table/Table.js (hunk 6 above)
Comment (petrican) [reply frontend/src/components/table/Table.js]: put `rowsPool` as it is shorter. Hope is fine

File: frontend/src/components/table/Table.js (hunk 7 above)
Comment (petrican) [reply frontend/src/components/table/Table.js]: put `multiSelectionStartIdx`

File: frontend/src/components/table/Table.js (hunk 3 above)
Comment (petrican) [reply frontend/src/components/table/Table.js]: do you mean     `this.multiSelectionStartIdx = this.multiSelectionStartIdx || currentId; ` ? This is not going to work.
Reason: https://www.loom.com/share/c46b280c4c8046fba0a0c0d537506e35

File: frontend/src/components/table/Table.js [hunk 8]
```diff
@@ -174,47 +193,67 @@ export default class Table extends PureComponent {
...
+              showSelectedIncludedView([rowsPool[currentId + 1]])
           );
+          this.clearMultiSelectionStartIdx();
         } else {
-          handleSelect(array[currentId + 1], false, idFocused);
+          this.updateMultiSelectionStartIdx(currentId);
+
+          const downShiftSel = rowsPool.slice(
+            this.multiSelectionStartIdx > 0 ? this.multiSelectionStartIdx : 0,
+            currentId + 2 // +2 because we want to slice up to the next row and include it
```
Comment (teosarca) [inline frontend/src/components/table/Table.js:217]: why also next row?
what's the reason behind?

Comment (petrican) [reply frontend/src/components/table/Table.js:217]: If you put 1 it will remain on the exact same row where it was all the time

**Comment by teosarca** [comment]
@petrican  lgtm

**Pull request merged**